File: 78Subsets.go
package main

import (
	"fmt"
)

func subsets(nums []int) [][]int {
	var output [][]int
	output = append(output, []int{})

	var curPath []int

	var backtrack func([]int)
	backtrack = func(curChoice []int) {
		for i := 0; i < len(curChoice); i++ {
			curPath = append(curPath, curChoice[i])
			tmp := make([]int, len(curPath))
			copy(tmp, curPath)
			output = append(output, tmp)
			backtrack(curChoice[i+1:])
			curPath = curPath[:len(curPath)-1]
		}
	}

	backtrack(nums)

	return output
}

func main() {
	output := subsets([]int{1, 2, 3})

	fmt.Println("\n******** output *********\n", output)
}

File: 78Subsets.cpp
#include <iostream>
#include <set>
#include <vector>
#include <iterator>
#include <utility>
#include <cstdlib>
#include <limits>
#include <functional>
#include <stack>
#include <unordered_set>
#include <queue>

using namespace std;

template<class Container>
void printContainer(const Container& container) {
//    return;

    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }

    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[";
        }
        cout << *it;
        if (it == prev(container.end())) {
            cout << "]" << endl;
        } else {
            cout << ", ";
        }
    }
}

template<class T>
void print2DimVec(const vector<vector<T>>& container) {
//    return;

    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }

    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[" << endl;
        }
        printContainer(*it);
        if (it == prev(container.end())) {
            cout << "]" << endl;
        }
    }
}

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> output;
        output.emplace_back(vector<int>{});
        vector<int> curPath;

        function<void(vector<int> &&)> backtrack =
            [&](vector<int>&& curChoices) {
                for (auto it = curChoices.begin(); it != curChoices.end();
                     it++) {
                    curPath.push_back(*it);
                    output.push_back(curPath);
                    backtrack(vector<int>(next(it), curChoices.end()));
                    curPath.pop_back();
                }
            };

        backtrack(move(nums));

        return output;
    }
};


int main(){
    Solution s;
    vector<int> input = vector<int>{1, 2, 3};
	auto output = s.subsets(input);

    cout << "*************output*************" << endl;
    print2DimVec(output);

    return 0;
}
